### ISSUE TITLE
Improving the execution of scripts (particularly on Windows)

### ISSUE THREAD

**Issue by bradenhilton** [twpayne/chezmoi#1238]
# I will preface all of this with the following:
- I am still new to chezmoi and could be wrong about many things below.
- My install is quite fresh at the moment so I may be missing any settings which would alleviate these issues.
- When I mention Windows, I am specifically referring to Windows and not WSL, which I assume does not suffer from these issues.
- I am aware that the testscript for `modify_` currently says Unix only with a FIXME comment.
- Apologies if there are any typos or mistakes below, I will try to fix them ASAP.

## Is your feature request related to a problem? Please describe.

To the best of my knowledge (and as you might already be aware) there is no way to execute any type of file on Windows (without any kind of interpreter) if it doesn't have a file extension. Currently this results in `exec` throwing an error like: `<path_to_some_file>: %1 is not a valid Win32 application` on script files.

I am aware that currently scripts must include a shebang or be executable binaries, this works on Unix-like systems because the files don't need extensions, and the shebang indicates which interpreter should be used.

I've noticed that there is a quick and dirty workaround to this problem - wrap a call to the required interpreter with the script inside a batch file* or similar e.g.:

```
# run_once_test.bat
python path\to\python_test.py
```

\* Presumably this is because my $SHELL is currently set to cmd.exe, so `exec` invokes .bat files through cmd.exe, which can run .bat files as-is?

Anyway, this is not optimal for a few reasons, including:
1. It defers to the system to find the interpreter (good I guess, assuming the interpreter is aliased or in $PATH) but it's an ugly hack (bad, but could be acceptable if this was the only issue).
2. Scripts already break chezmoi's declarative nature, and this breaks it even more by referencing and relying on a script which may not even exist.
3. It does not work as intended with `modify_` scripts, as the .bat extension is applied to the target file name (this gets more complicated below).

## Describe the solution you'd like

Something like the following:
1. Examine the extension of a script file and map the extension to one or more interpreters.
2. Search the user's $PATH for these interpreters (chezmoi already searches for other applications such as BitWarden so I think this is reasonable).
3. If none can be found, error out and inform the user to either ensure their path is configured correctly, or suggest to install one.
4. If an interpreter is found, call it in `exec` and run the script with it.

Care must be taken with `modify_` scripts to avoid the issue mentioned above, as the script must retain the correct extension in order to run, but the target file must not include this extension. I have attempted to illustrate this with an image:
![modify_script_1](https://user-images.githubusercontent.com/23157013/120099112-9c9c9d80-c131-11eb-86fb-75be4cee1320.png)

Let's examine the current outcome on an example file, `modify_settings.js`:
- On a Unix-like system, a new script is created named `<random>.settings.js`, and it modifies a file named `settings.js`. The script can be written in basically any language as the extension is not necessary (thanks to the use of a shebang).
- On a Windows system, a new script is created named `<random>.settings.js`, the script is run as if it were an executable binary, which is not possible, so it produces an error like the one I described above.

And now the Windows outcome on an example file `modify_settings.js.bat`:
- A new script is created named `<random>.settings.js.bat`, the new script runs successfully, and the target file is created with the name `settings.js.bat`.

We can make the script run, but it no longer affects the intended file.

If the primary file extension is consumed/stripped on Windows, `modify_settings.js` would affect different files on each platform. Maybe this means a change like this would need to affect Unix-like systems as well for parity, or the differences would need to be clearly documented.

It would get a little confusing if you intended to modify a file named `settings.js` with a script written in JavaScript. If the primary file extension is stripped, you'd have to name the script `modify_settings.js.js`.

The mappings don't need to be exhaustive (though surely there aren't too many languages and interpreters). Honestly, just CMD/PowerShell/Python alone are probably enough for like 90% of scripts on Windows. A good starting point can be found here: https://en.wikipedia.org/wiki/Cross-platform_software#Scripts_and_interpreted_languages.

## Describe alternatives you've considered

Allow the user to create their own .ext->interpreter mappings, possibly in the chezmoi config or `.chezmoidata`, maybe in a similar way to how `cd` is defined:

```
[.ps1]
    interpreter: ["powershell", "pwsh"]
    args: ["-NoLogo"]

[.py]
    interpreter: ["python"]
    args: []
```

Alternatively, a custom path could be supplied instead of an interpreter name. This way, you don't need to implement some sort of $PATH check internally, just match the extension and get the path from whichever file it's defined in. It's also possible that a user might want to use a specific interpreter (older or alternate version) without having to edit their $PATH.

---

There are of course other ways of determining which interpreter is needed. Perhaps you could mandate shebangs and parse them for the interpreter paths, though I'm not sure how that would interact with a Windows system when the script is run.

You could also use `assoc` and `ftype` to try to determine which interpreter should be used, though this requires the correct entry to be present in the registry, and some portable installs might not provide these entries.

In my opinion, using the file extension to determine which interpreter should be used is best because the extension is required regardless by default. It's possible to run `python <filename_no_extension>` where `<filename_no_extension>` is a valid Python script, but the trick is knowing that `<filename_no_extension>` is a Python script in the first place.

If you don't like the double file extension thing, maybe something like `modify_settings.js_with_python`? I personally don't like this, and have put no thought into its implementation, but I thought I'd just throw it out there.

## Additional context

I'm currently familiarising myself with the source code, but I don't know much about working on a piece of software of this size, and I know even less than that about Go.

Regardless, I've managed to throw together a sort of proof of concept of the mapping here:
https://gist.github.com/bradenhilton/71f490599b14ae269bdce79abf0aa6fa

Some (perhaps rambly) notes/findings:
- I have only tested this manually so far, so I haven't tested every combination imaginable. I didn't want to invest too much time in case you aren't even remotely interested in this.
- The file variable should be a valid path to an actual script file.
- `.ps1`/`.py` files: The test correctly identifies PowerShell and Python scripts, retrieves the paths of the interpreters from $PATH, and `exec`s them.
- `.rb` files: I don't have Ruby installed on this machine, and the test correctly identifies this and suggests I install it.

I'm leaning towards a "Windows only" label, but I suppose you could implement these changes on all platforms so I am still undecided. Obviously you may add any labels you deem appropriate.

Thank you for your time.

**Comment by twpayne** [comment]
Re-opening as this isn't quite all fixed yet.

I think this is a Windows-only issue (hence the `windows-only` label I just added). Reasoning: other OSes don't need this functionality as they already have a way to associate scripts with arbitrary interpreters (i.e. `#!` lines).

> You could also use `assoc` and `ftype` to try to determine which interpreter should be used, though this requires the correct entry to be present in the registry, and some portable installs might not provide these entries.

These could provide more sensible defaults than the current hard-coded defaults introduced in #1239. Do you think this is worthwhile?

**Comment by bradenhilton** [comment]
> These could provide more sensible defaults than the current hard-coded defaults introduced in #1239. Do you think this is worthwhile?

It's tough to say to be honest, sometimes the association is for something other than an interpreter (such as an editor), though I suppose you could just override this in the config. I'd maybe keep PowerShell hardcoded at a minimum, as `ftype` shows that `.ps1` scripts are set to open in `notepad.exe` by default, and PowerShell is pre-installed on basically every version of Windows.

Perhaps it's worth cc'ing @zb140 also.

**Comment by zb140** [comment]
Yes, Powershell is deliberately (and disappointingly) set to open with `notepad.exe` by default as a "security" measure. I would certainly be in favor of providing a default mapping for `.ps1`.

It's true that the `assoc` and `ftype` settings could be set to open scripts in something other than an interpreter, but then "running" them from the command line with `.\script.py` (for example) will also just open the editor.  My thought was to extract the associations for only the extensions in the the `PATHEXT` environment variable. Those can reasonably be presumed to be things the user expects to be able to invoke as commands. That's what the script I mentioned in the pull request comments does. I can elaborate further a little later this evening but I've stepped away from my computer for a while and I don't feel like typing anything longer on my phone 😁

**Comment by zb140** [comment]
### Context
Feel free to TL;DR -- I'll put my conclusion at the end :smile:

The `PATHEXT` environment variable is a list of all of the file extensions that should be considered executable.  This is what allows you to run a program as (for example) `chezmoi` rather than `chezmoi.exe`.  `assoc` is a shell built-in in `cmd.exe` that associates extensions with arbitrary "file types" that are conceptually a bit like mime types.  `ftype` is another built-in that associates one of those file types with a command to be run to "open" the file.  This could mean either execute it (if it's a program) or open it in its program (if it's data).  Of course, a script file lives in the weird middle ground.

Windows also has multiple different APIs for executing another program.  `CreateProcess`, which is what's used by Go's `os/exec`, is pretty close in concept to a Unix-style `exec`.  It doesn't do much of anything smart; it just assumes the filename you've given it is executable and tries to launch it.  It doesn't take _any_ of the file type stuff into account.  Then there's `ShellExecute`, which _does_ take it into account.  One of its arguments is a "verb"; the verb "open" is what's configured by the `ftype` command I mentioned. `open` is the default verb for most file types, though it is possible to change the default for individual types.

If you try to execute a Perl script, for example, in a shell like this:
```
C:\> .\script.pl
```
and you have not set up the `assoc` and `ftype` mappings appropriately, it will throw an error saying that that file isn't a command it can execute.  Likewise, if you try to execute `script.pl` like this:
```
C:\> .\script
```
and `.pl` is not in your `PATHEXT`, it will fail, even if you _have_ set up the file type associations correctly.

### TL;DR? Here's my thoughts

I think it's reasonable to assume that extensions listed in `PATHEXT` are ones that users will expect to be able to execute as scripts, and conversely, extensions that are _not_ in `PATHEXT` are not likely to be executable directly from a shell

**Comment by zb140** [comment]
Ugh, a github keyboard shortcut "helped" me by closing this 😦 I'll edit my above comment incomplete comment to finish my thoughts.

**Comment by bradenhilton** [comment]
@zb140 I've only had a cursory look through the script at the moment, but does this look like the intended/expected output to you?

```
> $env:PATHEXT
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW;.CPL

> .\interpreters.ps1 -Format toml
[interpreters.msc]
command = 'C:\WINDOWS\system32\mmc.exe'

```

**Comment by zb140** [comment]
@bradenhilton Not quite, but close -- I would expect to see python in there too.

Here's mine, for example:
```
> $env:PATHEXT
.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL;.PS1;.PY;.PYW

> .\interpreters.ps1 -Format toml
[interpreters.py]
command = 'C:\Windows\py.exe'

[interpreters.msc]
command = 'C:\Windows\system32\mmc.exe'

[interpreters.pyw]
command = 'C:\Windows\pyw.exe'

[interpreters.ps1]
command = 'C:\Program Files\PowerShell\7\pwsh.exe'
```

By default, the script will also exclude any extension whose interpreter is `wscript.exe` (the [Windows Script Host](https://en.wikipedia.org/wiki/Windows_Script_Host)).  I made it the default mostly because it "hijacks" some common extensions -- most notably, it assumes `.js` files are actually [JScript](https://en.wikipedia.org/wiki/JScript) using WSH features.

Could you try it like this? This will disable the exclusion and hopefully we can see why python isn't showing up for you.
```
> .\interpreters.ps1 -Format toml -ExcludeCommand @()
```

**Comment by bradenhilton** [comment]
@zb140 If I had to guess I'd say it might be because I installed Python via [Scoop](https://scoop.sh/). ([manifest](https://github.com/ScoopInstaller/Main/blob/master/bucket/python.json))

Otherwise I'm working on a relatively fresh (days old, barely used) Windows install.

```
❯ .\interpreters.ps1 -Format toml -ExcludeCommand @()
[interpreters.wsh]
command = 'C:\WINDOWS\System32\WScript.exe'

[interpreters.js]
command = 'C:\Windows\System32\WScript.exe'

[interpreters.msc]
command = 'C:\WINDOWS\system32\mmc.exe'

[interpreters.vbe]
command = 'C:\WINDOWS\System32\WScript.exe'

[interpreters.jse]
command = 'C:\Windows\System32\WScript.exe'

[interpreters.wsf]
command = 'C:\WINDOWS\System32\WScript.exe'

[interpreters.vbs]
command = 'C:\WINDOWS\System32\WScript.exe'

```

It's not a big deal to me, I think I'd rather use the interpreter name as the command and get it from my $PATH, but if true it might be worth considering as there are several methods of installing software on Windows (direct download/Chocolatey/Scoop/winget etc).
